torch_geometric/nn/models/rbcd_attack.py

```diff
@@ -0,0 +1,760 @@
+
+        scalars['metric'] = metric.item()
+        return scalars
+
+    @torch.no_grad()
+    def _close(self, x: Tensor, labels: Tensor, budget: int,
+               idx_attack: Optional[Tensor] = None,
+               **kwargs) -> Tuple[Tensor, Tensor]:
+        """Clean up and prepare return argument."""
+        # Retrieve best epoch if early stopping is active
+        # (not explicitly covered by pseudo code)
+        if self.coeffs['with_early_stopping']:
+            self.current_block = self.best_block.to(self.device)
+            self.block_edge_index = self.best_edge_index.to(self.device)
+            self.block_edge_weight = self.best_pert_edge_weight.to(self.device)
+
+        # Sample final discrete graph (Algorithm 1, line 16)
+        edge_index, flipped_edges = self._sample_final_edges(
+            x, labels, budget, idx_attack=idx_attack, **kwargs)
+
+        return edge_index, flipped_edges
+
+    def _forward(self, x: Tensor, edge_index: Tensor, edge_weight: Tensor,
+                 **kwargs) -> Tensor:
+        """Forward model."""
+        return self.model(x, edge_index, edge_weight, **kwargs)
+
+    def _forward_and_gradient(self, x: Tensor, labels: Tensor,
+                              idx_attack: Optional[Tensor] = None,
+                              **kwargs) -> Tuple[Tensor, Tensor]:
+        """Forward and update edge weights."""
+        self.block_edge_weight.requires_grad = True
+
+        # Retrieve sparse perturbed adjacency matrix `A \oplus p_{t-1}`
+        # (Algorithm 1, line 6 / Algorithm 2, line 7)
+        edge_index, edge_weight = self._get_modified_adj(
+            self.edge_index, self.edge_weight, self.block_edge_index,
+            self.block_edge_weight)
+
+        # Get prediction (Algorithm 1, line 6 / Algorithm 2, line 7)
+        prediction = self._forward(x, edge_index, edge_weight, **kwargs)
+        # Calculate loss combining all each node
+        # (Algorithm 1, line 7 / Algorithm 2, line 8)
+        loss = self.loss(prediction, labels, idx_attack)
+        # Retrieve gradient towards the current block
+        # (Algorithm 1, line 7 / Algorithm 2, line 8)
+        gradient = torch.autograd.grad(loss, self.block_edge_weight)[0]
+
+        return loss, gradient
+
+    def _get_modified_adj(self, edge_index: Tensor, edge_weight: Tensor,
+                          block_edge_index: Tensor,
+                          block_edge_weight: Tensor) -> Tuple[Tensor, Tensor]:
+        """Merges adjacency matrix with current block (incl. weights)"""
+        if self.is_undirected_graph:
+            block_edge_index, block_edge_weight = to_undirected(
+                block_edge_index, block_edge_weight, num_nodes=self.num_nodes,
+                reduce='mean')
+
+        modified_edge_index = torch.cat(
+            (edge_index.to(self.device), block_edge_index), dim=-1)
+        modified_edge_weight = torch.cat(
+            (edge_weight.to(self.device), block_edge_weight))
```

Comment:
As mentioned above, would be great to get rid of these device transfers.

Reply:
See last week's high-level comment